base/reduce.jl

```diff
@@ -362,10 +367,105 @@ julia> reduce(*, [2; 3; 4]; init=-1)
 -24
 ```
 """
-reduce(op, itr; kw...) = mapreduce(identity, op, itr; kw...)
+function reduce(op, itr::T; kw...) where T
+    # Redispatch, adding traits
+    reduce(op, itr, eltype_or_default_eltype(itr), IteratorSize(T); kw...)
+end
+
+function reduce(op, itr, et, isize; kw...)
+    # Fallback: if nothing interesting is being done with the traits
+    # or the operation
+    return mapreduce(identity, op, itr; kw...)
+end
 
 reduce(op, a::Number) = a  # Do we want this?
 
+##### Operation specific reduce optimisations
+
+## vcat
+
+function reduce(::typeof(vcat), xs, T::Type{<:AbstractVector}, isize)
```

Comment:
This should probably be restricted to the case where all vectors are of the same concrete type. Otherwise there is no guaranty that calling `vcat` repeatedly (what `reduce` does by default) will return a vector of the same type as the first one.